core/src/main/java/io/grpc/ServiceProviders.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+final class ServiceProviders {
+  private static final Logger logger = Logger.getLogger(ServiceProviders.class.getName());
+
+  private ServiceProviders() {
+    // do not instantiate
+  }
+
+  /**
+   * If this is not Android, returns the highest priority implementation of the class via
+   * {@link ServiceLoader}.
+   * If this is Android, returns an instance of the highest priority class in {@code hardcoded}.
+   */
+  public static <T extends ServiceProvider> T load(
+      Class<T> klass, List<Class<?>> hardcoded, ClassLoader cl) {
+    List<T> candidates = loadAll(klass, hardcoded, cl);
+    if (candidates.isEmpty()) {
+      return null;
+    }
+    return candidates.get(0);
+  }
+
+  /**
+   * If this is not Android, returns all available implementations discovered via
+   * {@link ServiceLoader}.
+   * If this is Android, returns all available implementations in {@code hardcoded}.
+   * The list is sorted in descending priority order.
+   *
+   * <p>If a failure was encountered while initializing any class, then the result is empty.
+   */
+  public static <T extends ServiceProvider> List<T> loadAll(
+      Class<T> klass, List<Class<?>> hardcoded, ClassLoader cl) {
+    Iterable<T> candidates;
+    if (isAndroid(cl)) {
+      candidates = getCandidatesViaHardCoded(klass, hardcoded);
+    } else {
+      candidates = getCandidatesViaServiceLoader(klass, cl);
+    }
+    List<T> list = new ArrayList<T>();
+    Iterator<T> iter = candidates.iterator();
+    try {
+      while (iter.hasNext()) {
+        T current = iter.next();
+        if (!current.isAvailable()) {
+          continue;
+        }
+        list.add(current);
+      }
+
+      // Sort descending based on priority.
+      Collections.sort(list, Collections.reverseOrder(new Comparator<ServiceProvider>() {
+        @Override
+        public int compare(ServiceProvider f1, ServiceProvider f2) {
+          return f1.priority() - f2.priority();
+        }
+      }));
+      return Collections.unmodifiableList(list);
+    } catch (Throwable t) {
```

Comment:
Can we please not have unnecessary behavior changes in a refactoring change like this? This can be done easily as a follow-up and avoids it being lost in the noise and conflated in review.

Reply:
Updated PR with this error handling code removed